### DISCUSSION TITLE
How to reliably resubscribe when remounting an atom?

### DISCUSSION THREAD

**Discussion by Pinpickle** [pmndrs/jotai#2468] [Q&A]
If an atom is subscribing to a value where the current value can always be known (e.g. the time), you can use `onMount` to update it when the value changes:

```ts
const nowToTheSecond = () => {
  return Math.floor(performance.now() / 1000);
};

const timerAtom = atom(nowToTheSecond());
timerAtom.onMount = (setSelf) => {
  setSelf(nowToTheSecond());
  const interval = setInterval(() => {
    setSelf(nowToTheSecond());
  }, 1000);

  return () => {
    clearInterval(interval);
  };
};
```

In the above example, there's a re-render cycle when remounting this atom where the previous value is used. I understand that this is because `onMount` will be called in a `useEffect` (or something akin to that).

If I log the value of the atom in a `useEffect`:

```tsx
const Timer = () => {
  const time = useAtomValue(timerAtom);

  useEffect(() => {
    const current = nowToTheSecond();
    console.log(
      `Current: ${current}\natom: ${time}\ndifference: ${Math.abs(
        current - time
      )}`
    );
  }, [time]);

  return <div>The time is: {time}</div>;
};
```

Then I can see the timer out of sync when re-mounting it (the duplicate log of the incorrect one is due to strict mode):

```
Current: 568
atom: 568
difference: 0

Disabling timer

Enabling timer

Current: 575
atom: 568
difference: 7

Current: 575
atom: 568
difference: 7

Current: 575
atom: 575
difference: 0
```

I'm aware that there are ways to work around this - e.g. making sure the atom stays mounted. I'm checking to see if there's a way to make sure the value is correct if it is unmounted and remounted. Like I could do if I was invoking `useExternalStore` directly.

See it for yourself here: https://stackblitz.com/edit/vitejs-vite-h3vbjd?file=src%2FApp.tsx

**Comment by dai-shi** [comment]
I think it's working as intended. Our design decision is to keep previous values while atoms are unmounted, until they are garbage collected.

onMount and onUnmount should be paired. So, avoiding keeping the previous value might feel more intuitive.
```ts
const timerAtom = atom<number | undefined>(undefined);
timerAtom.onMount = (setSelf) => {
  setSelf(nowToTheSecond());
  const interval = setInterval(() => {
    setSelf(nowToTheSecond());
  }, 1000);

  return () => {
    clearInterval(interval);
    setSelf(undefined);
  };
};
```

**Reply by Pinpickle** [reply]
I hadn't considered un-setting on unmount! This seems reasonable given the constraints.

It does feel like a hack, though. There isn't a technical reason we can't do this in pure React, for example.

It would be great to have something that lets us read the most up-to-date value on mount _before_ mounting. One way to implement this is to have the `read` function called every time when an atom is unmounted and it's only cached when an atom is mounted. 

This wouldn't affect primitive atoms with a set value. They would keep their value across unmounts as they do already.

**Reply by dai-shi** [reply]
> There isn't a technical reason we can't do this in pure React, for example.

What would you do with pure React and a global variable?

**Reply by Pinpickle** [reply]
@dai-shi this is a pure React implementation of the same timer using `useSyncExternalStore`. It doesn't have the issue of an out-of-sync first render.

```tsx
class TimerObject {
  getValue = () => {
    return nowToTheSecond();
  };

  subscribe = (cb: () => void) => {
    const interval = setInterval(() => {
      cb();
    }, 1000);

    return () => {
      clearInterval(interval);
    };
  };
}

const globalTimer = new TimerObject();

const PureReactTimer = () => {
  const time = useSyncExternalStore(
    globalTimer.subscribe,
    globalTimer.getValue
  );

  useEffect(() => {
    const current = nowToTheSecond();
    console.log(
      `Current: ${current}\nstate: ${time}\ndifference: ${Math.abs(
        current - time
      )}`
    );
  }, [time]);

  return <div>The time is: {time}</div>;
};
```

**Reply by dai-shi** [reply]
The equivalent in Jotai is this:
```js
const refreshAtom = atom(0);
refreshAtom.onMount = (set) => {
  const id = setInterval(() => set((c) => c + 1), 1000);
  return () => clearInterval(id);
};
const timerAtom = atom((get) => {
  get(refreshAtom);
  return nowToTheSecond();
});
```
Well, not really equivalent, but it's close.

**Reply by Pinpickle** [reply]
> The equivalent in Jotai is this:

@dai-shi this solution has the same potential for an out-of-sync first render, right?

**Reply by dai-shi** [reply]
Oh, yeah, because we cache the atom value (otherwise, it doesn't work for promises.)

I don't think what you want is possible with "global" state.
Just a simpler hack:
```js
const refreshAtom = atom(0);
refreshAtom.onMount = (set) => {
  const id = setInterval(() => set((c) => c + 1), 1000);
  return () => clearInterval(id);
};
const useTimer = () => {
  useAtom(refreshAtom);
  return nowToTheSecond();
};
```

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/discussions/2468#discussioncomment-8911110
A working example with `undefined`:
https://stackblitz.com/edit/vitejs-vite-j4vync?file=src%2FApp.tsx

**Reply by Pinpickle** [reply]
@dai-shi this is great!

What happens if a suspense is triggered so the `timerAtom` is never mounted (and therefore never unmounted and not reset)? Will there be an out-of-sync value?

**Reply by Pinpickle** [reply]
Suspense causes the bug to resurface: https://stackblitz.com/edit/vitejs-vite-m3zu2h?file=src%2FApp.tsx

Especially so if you disable the timer while the component is suspending.

```
Current: 2154
atom: 2154
difference: 0
Disabling timer
Enabling timer
Disabling timer
Enabling timer
Current: 2336
atom: 2274
difference: 62
```

**Reply by dai-shi** [reply]
It will still be `undefined`, right? It can be out of sync, if `useEffect` is delayed. https://github.com/pmndrs/jotai/discussions/2468#discussioncomment-8968386 can avoid it.
